### PR TITLE
PrimitiveState: added Generic Constraints

### PR DESCRIPTION
**Pull request by UseMuse**

### CODE REVIEW

File: projects/library/src/lib/states/primitive.state.ts [hunk 1]
```diff
@@ -4,4 +4,6 @@ import { BaseState } from './base.state';
  *	@class
  *	@classdes Primitive state class. Used for save state with Primitive type (Like boolean, number or string).
  */
-export class PrimitiveState<T> extends BaseState<T> {}
+export class PrimitiveState<
+  T = string | number | boolean | bigint | symbol
```
Comment (dreyliky) [inline projects/library/src/lib/states/primitive.state.ts:8]: This approach isn't flexible. What if I want to use some concrete `type` or `key of enum` as the primitive?

Like:
```
enum MyEnum {
   Value1,
   Value2
}

type MyType = 'male' | 'female';

PrimitiveState<MyEnum>
PrimitiveState<MyType>
```

**Review by dreyliky — CHANGES_REQUESTED** [review]

File: projects/library/src/lib/states/primitive.state.ts (hunk 1 above)
Comment (UseMuse) [reply projects/library/src/lib/states/primitive.state.ts:8]: how about using ObjectState for such cases and leaving primitive for explicit primitives?

Comment (UseMuse) [reply projects/library/src/lib/states/primitive.state.ts:8]: or, in order to cover all areas, make EnumState and TypeState

Comment (UseMuse) [reply projects/library/src/lib/states/primitive.state.ts:8]: although everything is correct, set is typed T and maybe my PR is overkill

Comment (UseMuse) [reply projects/library/src/lib/states/primitive.state.ts:8]: on the other hand without T = string | number | boolean | bigint | symbol nothing prevents us from using PrimitiveState (and other implementations of BaseState) with any type T, for example 
``` export class SelectedUser extends PrimitiveState<{name: 'name', last:'last'}> {} ```

File: projects/library/src/lib/states/array.state.ts [hunk 2]
```diff
@@ -7,217 +7,235 @@ import { BaseState } from './base.state';
...
+    this.setNewValue(items);
+  }
+
+  /**
+   * 	Push item to array in state if not exists in state
+   *	@public
+   *	@param {Generic} item - Item needs to push
+   */
+  @Action
+  public pushItemIfNotExists(item: T): void {
```
Comment (dreyliky) [inline projects/library/src/lib/states/array.state.ts:96]: It looks like some kind of custom action to me.

1) If you want to implement this logic in your specific state, you can create custom action like:

```
@NgxState()
@Injectable()
export class TodoItemsState extends ArrayState<TodoItem[]> {
    public pushIfNotExists(todoItem: TodoItem): void {
        if (
            this.data &&
            this.data.filter((i) => this.compareItems(i, item)).length === 0
        ) {
            this.pushItem(item);
        }
    }
}
```

2) If you want to use this logic often in your project, you can create your own `base state`, like:

```
// Abstract base state
@NgxState()
export abstract class ArrayWithUniqueItemsState<T> extends ArrayState<T> {
    public override pushItem(item: T): void {
        if (
            this.data &&
            this.data.filter((i) => this.compareItems(i, item)).length === 0
        ) {
            super.pushItem(item);
        }
    }
    
    ...
}

// Concrete state
@NgxState()
@Injectable()
export class TodoItemsState extends ArrayWithUniqueItemsState<TodoItem> {}
```

**Comment by dreyliky** [comment]
PrimitiveState features implemented in dev branch

**Pull request closed without merging**